src/core/stream/representation/downloading_queue.ts

```diff
@@ -0,0 +1,510 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  BehaviorSubject,
+  concat as observableConcat,
+  defer as observableDefer,
+  Observable,
+  of as observableOf,
+  merge as observableMerge,
+  EMPTY,
+  ReplaySubject,
+} from "rxjs";
+import {
+  filter,
+  finalize,
+  map,
+  mergeMap,
+  share,
+  switchMap,
+  take,
+} from "rxjs/operators";
+import { ICustomError } from "../../../errors";
+import log from "../../../log";
+import Manifest, {
+  Adaptation,
+  ISegment,
+  Period,
+  Representation,
+} from "../../../manifest";
+import {
+  ISegmentParserParsedInitSegment,
+  ISegmentParserParsedSegment,
+} from "../../../transports";
+import assert from "../../../utils/assert";
+import assertUnreachable from "../../../utils/assert_unreachable";
+import objectAssign from "../../../utils/object_assign";
+import {
+  IPrioritizedSegmentFetcher,
+  IPrioritizedSegmentFetcherEvent,
+} from "../../fetchers";
+import {
+  IQueuedSegment,
+} from "../types";
+
+/**
+ * Class scheduling segment downloads for a single Representation.
+ * @class DownloadingQueue
+ */
+export default class DownloadingQueue<T> {
+  /** Context of the Representation that will be loaded through this DownloadingQueue. */
+  private _content : IDownloadingQueueContext;
+  /**
+   * Observable doing segment requests and emitting related events.
+   * We only can have maximum one at a time.
+   * `null` when `start` has never been called.
+   */
+  private _currentObs$ : Observable<IDownloadingQueueEvent<T>> | null;
+  /**
+   * Current queue of segments scheduled for download.
+   *
+   * Segments whose request are still pending are still in that queue. Segments
+   * are only removed from it once their request has finished.
```

Comment:
"has finished" does it also wrap when the request is in Error?

Reply:
The queue is completely interrupted if the request at that step fails (retries are peformed at the fetchers' level so here this is abstracted and we're only left with real errors).

So you're right that "has finished" might not be the best term. Maybe "has succeeded"?

Reply:
ha ok, cool, I think yes it could bring something by saying only "has succeeded" :)

Reply:
done